src/drivers/mmc/host/sdhci/sdhci_ops.c

```diff
@@ -0,0 +1,195 @@
+	uint32_t wcmd = 0;
+	int retry = RETRIES;
+
+	wcmd |= (cmd_idx & 0x3f) << 24;
+
+	if (!(flags & MMC_RSP_PRESENT)) {
+	} else if (flags & MMC_RSP_136) {
+		wcmd |= 0x1 << 16;
+	} else if (flags & MMC_RSP_BUSY){
+		wcmd |= 0x3 << 16;
+	} else {
+		wcmd |= 0x2 << 16;
+	}
+
+	if (flags & MMC_RSP_CRC) {
+		wcmd |= 1 << 19;
+	}
+
+	if (flags & MMC_DATA_XFER) {
+		wcmd |= 0x1 << 21;
+		/* enable DMA in imx it places in MIX_CTRL */
+		wcmd |= 0x1;
+		if (flags & MMC_DATA_READ) {
+			wcmd |= 0x10;
+		}
+	}
+
+	sdhci_writel(host, USDHC_CMD_ARG, arg);
+
+	log_debug("send cmd: 0x%08x idx = %d arg = %x", wcmd, cmd_idx, arg);
+
+	sdhci_writel(host, USDHC_CMD_XFR_TYP, wcmd);
+
+	/* Wait command to be completed */
+	while(!(sdhci_readl(host, USDHC_INT_STATUS) & 0x1) && retry > 0) {
+		usleep(USEC_PER_MSEC);
+		retry--;
+	}
+
+	if (retry == 0) {
+		log_debug("Timeout!");
+	}
+
+	log_debug("USDHC_INT_STATUS=%08x", sdhci_readl(host, USDHC_INT_STATUS));
+	sdhci_writel(host, USDHC_INT_STATUS, 0x1);
+}
+
+static void sdhci_reset(struct sdhci_host *host) {
+	sdhci_writel(host, USDHC_SYS_CTRL, sdhci_readl(host, USDHC_SYS_CTRL) | USDHC_SYS_CTRL_RSTA);
+
+	int timeout = TIMEOUT;
+	while(timeout--);
+
+	if (sdhci_readl(host, USDHC_SYS_CTRL) & USDHC_SYS_CTRL_RSTA) {
+		log_error("Reset timeout");
+	}
+
+	sdhci_writel(host, USDHC_VEND_SPEC,
+			sdhci_readl(host, USDHC_VEND_SPEC) | (1 << 31) | (1 << 14) | (1 << 13));
+}
+
+static int sdhci_get_cd(struct mmc_host *host) {
+	assert(host);
+
+	if (sdhci_readl(host->priv, USDHC_PRES_STATE) & USDHC_PRES_STATE_CDPL) {
+		return 1;
+	} else {
+		return 0;
+	}
+}
+
+static void sdhci_mmc_request(struct mmc_host *mmc_host, struct mmc_request *req) {
+	struct sdhci_host *host;
+
+	assert(mmc_host);
+	assert(req);
+
+	host = mmc_host->priv;
+
+	log_debug("req->cmd.opcode %x, req->cmd.arg %x, req->cmd.flags %x", req->cmd.opcode, req->cmd.arg, req->cmd.flags);
+
+	if (req->cmd.flags & MMC_DATA_XFER) {
+		sdhci_writel(host, USDHC_INT_STATUS, -1);
+
+		sdhci_writel(host, USDHC_DS_ADDR, (uint32_t) req->data.addr);
+	}
+
+	imx6_usdhc_send_cmd(host, req->cmd.opcode, req->cmd.arg, req->cmd.flags);
+
+	sdhci_cmd_done(host, req);
+}
+
+static const struct mmc_host_ops sdhci_mmc_ops = {
+	.request = sdhci_mmc_request,
+	.get_cd  = sdhci_get_cd,
+};
+
+int sdhci_hw_probe(struct sdhci_host *host) {
+	uint32_t tmp;
+	struct mmc_host *mmc;
+
+	sdhci_reset(host);
+
+	sdhci_writel(host, USDHC_SYS_CTRL, 0xb011f);
+	sdhci_writel(host, USDHC_INT_STATUS_EN, -1);
+	sdhci_writel(host, USDHC_PROT_CTRL, sdhci_readl(host, USDHC_PROT_CTRL) | 1 | 4);
+
+	tmp = 512 | (1 << 16);
+	sdhci_writel(host, USDHC_BLK_ATT, tmp);
+	if (sdhci_readl(host, USDHC_PRES_STATE) & USDHC_PRES_STATE_CDPL) {
+		mmc = mmc_alloc_host();
+		mmc->ops = &sdhci_mmc_ops;
+		mmc->priv = host;
+		mmc_sw_reset(mmc);
+		sdhci_reg_dump(host);
+	} else {
+		log_debug("CDPL bit is zero! %08x", sdhci_readl(host, USDHC_PRES_STATE));
```

Comment:
Is this situation properly handled? The upper layer will get 0 from this function, but mmc will be not allocated, it's OK?

Reply:
In this case there's no card inserted, so yes, it's ok not to alloc MMC